server/catgenome/src/main/java/com/epam/catgenome/util/MotifSearcher.java

```diff
@@ -0,0 +1,197 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.epam.catgenome.util;
+
+import com.epam.catgenome.component.MessageHelper;
+import com.epam.catgenome.constant.MessagesConstants;
+import com.epam.catgenome.entity.reference.motif.Motif;
+import com.epam.catgenome.manager.gene.parser.StrandSerializable;
+
+import java.util.Spliterators;
+import java.util.Spliterator;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public final class MotifSearcher {
+
+    private static final int NEGATIVE_STRAND = 1;
+    private static final byte CAPITAL_A = 'A';
+    private static final byte CAPITAL_C = 'C';
+    private static final byte CAPITAL_G = 'G';
+    private static final byte CAPITAL_T = 'T';
+    private static final byte CAPITAL_N = 'N';
+    private static final byte LOWERCASE_A = 'a';
+    private static final byte LOWERCASE_C = 'c';
+    private static final byte LOWERCASE_G = 'g';
+    private static final byte LOWERCASE_T = 't';
+    private static final byte LOWERCASE_N = 'n';
+    private static final String EMPTY_VALUE = "";
+
+    private MotifSearcher() {
+    }
+
+    public static List<Motif> search(final byte[] seq, final String regex, final String contig) {
+        return search(seq, regex, null, contig);
+    }
+
+    public static List<Motif> search(final byte[] seq, final String regex,
+                                     final StrandSerializable strand, final String contig) {
+        final String sequence;
+        final String negativeSequence;
+
+        if(strand == null) {
+            sequence = new String(seq);
+            negativeSequence = reverseAndComplement(seq);
+        } else if (strand == StrandSerializable.POSITIVE) {
+            sequence = new String(seq);
+            negativeSequence = EMPTY_VALUE;
+        } else if (strand == StrandSerializable.NEGATIVE) {
+            sequence = EMPTY_VALUE;
+            negativeSequence = reverseAndComplement(seq);
+        } else {
+            throw new IllegalArgumentException(MessageHelper.getMessage(MessagesConstants.ERROR_UNSUPPORTED_OPERATION,
+                    strand));
+        }
+
+        final Pattern pattern = Pattern.compile(convertIupacToRegex(regex), Pattern.CASE_INSENSITIVE);
```

Comment:
how about to configure matchers in condition on `strand` and it this case we will be able to create iterator only on one Mathcer, and do not do any redundant checks when itarating